component/loki/source/kubernetes/kubetail/tailer.go

```diff
@@ -0,0 +1,323 @@
+		}
+	}
+}
+
+// containerTerminated determines whether the container this tailer was
+// watching has terminated and won't restart. If containerTerminated returns
+// true, it means that no more logs will appear for the watched target.
+func (t *tailer) containerTerminated(ctx context.Context) (terminated bool, err error) {
+	var (
+		key           = t.target.NamespacedName()
+		containerName = t.target.ContainerName()
+	)
+
+	podInfo, err := t.opts.Client.CoreV1().Pods(key.Namespace).Get(ctx, key.Name, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+
+	// The pod UID is different than the one we were tailing; our UID has
+	// terminated.
+	if podInfo.GetUID() != kubetypes.UID(t.target.UID()) {
+		return true, nil
+	}
+
+	containerInfo, containerType, found := findContainerStatus(podInfo, containerName)
+	if !found {
+		return false, fmt.Errorf("could not find container %q in pod status", containerName)
+	}
+
+	restartPolicy := podInfo.Spec.RestartPolicy
+
+	switch containerType {
+	case containerTypeApp:
+		// An app container will only restart if:
+		//
+		// * It is in a waiting (meaning it's waiting to run) or running state
+		//   (meaning it already restarted before we had a chance to check)
+		// * It terminated with any exit code and restartPolicy is Always
+		// * It terminated with non-zero exit code and restartPolicy is not Never
+		//
+		// https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy
+		switch {
+		case containerInfo.State.Waiting != nil || containerInfo.State.Running != nil:
+			return false, nil // Container will restart
+		case restartPolicy == corev1.RestartPolicyAlways:
+			return false, nil // Container will restart
+		case containerInfo.State.Terminated != nil && restartPolicy == corev1.RestartPolicyOnFailure && containerInfo.State.Terminated.ExitCode != 0:
+			return false, nil // Container will restart
+		default:
+			return true, nil // Container will *not* restart
+		}
+
+	case containerTypeInit:
+		// An init container will only restart if:
+		//
+		// * It is in a waiting (meaning it's waiting to run) or running state
+		//   (meaning it already restarted before we had a chance to check)
+		// * It terminated with an exit code of non-zero and restartPolicy is not
+		//   Never.
+		//
+		// https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#understanding-init-containers
+		switch {
+		case containerInfo.State.Waiting != nil || containerInfo.State.Running != nil:
+			return false, nil // Container will restart
+		case containerInfo.State.Terminated != nil && containerInfo.State.Terminated.ExitCode != 0 && restartPolicy != corev1.RestartPolicyNever:
+			return false, nil // Container will restart
+		default:
+			return true, nil // Container will *not* restart
+		}
+
+	case containerTypeEphemeral:
+		// Ephemeral containers never restart.
+		//
+		// https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/
+		return true, nil
```

Comment:
Does this mean we'll never tail ephemeral containers though? Should we at least check if they're running?

Reply:
Yeah, good point (though this line is only hit after we got an EOF, so it would at least be tailed for a little while)